packages/ketcher-react/src/script/editor/shared/closest.js

```diff
@@ -505,21 +506,63 @@ function findCloseMerge(restruct, selected, maps = ['atoms', 'bonds'], scale) {
     )
   })
 
-  const result = {}
-  maps.forEach((mp) => {
-    result[mp] = Array.from(pos[mp].keys()).reduce((res, srcId) => {
-      const skip = { map: mp, id: srcId }
-      const item = findMaps[mp](restruct, pos[mp].get(srcId), skip, null, scale)
-
-      if (item && !selected[mp].includes(item.id)) res.set(srcId, item.id)
+  const result = {
+    atoms: new Map(),
+    atomToFunctionalGroup: new Map()
+  }
 
-      return res
-    }, new Map())
+  maps.forEach((map) => {
+    if (map === 'atoms') {
+      Array.from(pos.atoms.keys()).forEach((atomId) => {
+        const atomPosition = pos.atoms.get(atomId)
+        mergeAtomToAtom(atomId, restruct, atomPosition, selected, result) ||
+          mergeAtomToFunctionalGroup(atomId, restruct, atomPosition, result)
+      })
+    } else {
+      result[map] = Array.from(pos[map].keys()).reduce((res, srcId) => {
+        const skip = { map, id: srcId }
+        const item = findMaps[map](
+          restruct,
+          pos[map].get(srcId),
+          skip,
+          null,
+          scale
+        )
+
+        if (item && !selected[map].includes(item.id)) res.set(srcId, item.id)
+
+        return res
+      }, new Map())
+    }
   })
 
   return result
 }
 
+function mergeAtomToAtom(atomId, restruct, atomPosition, selected, result) {
+  const skip = { map: 'atoms', id: atomId }
+  const closestAtom = findClosestAtom(restruct, atomPosition, skip, null)
+
+  if (closestAtom && !selected.atoms.includes(closestAtom.id)) {
+    result.atoms.set(atomId, closestAtom.id)
+    return true
+  }
+
+  return false
+}
+
+function mergeAtomToFunctionalGroup(atomId, restruct, atomPosition, result) {
+  if (FunctionalGroup.isAttachmentPointAtom(atomId)) return false
+
+  const closestFunctionalGroup = findClosestFG(restruct, atomPosition, null)
+  if (closestFunctionalGroup) {
+    result.atomToFunctionalGroup.set(atomId, closestFunctionalGroup.id)
+    return true
+  }
+
+  return false
```

Comment:
Can we somehow join these conditions into single condition?

``` 
if (FunctionalGroup.isAttachmentPointAtom(atomId)) return false

return false
```

Reply:
FunctionalGroup.isAttachmentPointAtom(atomId) has already return you boolean

Maybe just like this 

`
  const closestFunctionalGroup = findClosestFG(restruct, atomPosition, null)
  if (closestFunctionalGroup) {
    result.atomToFunctionalGroup.set(atomId, closestFunctionalGroup.id)
    return true
  }

 return !FunctionalGroup.isAttachmentPointAtom(atomId)
`

Reply:
Thanks for your review again, but I can't think of a better way to join them into a single condition, and unfortunately, your suggestion is not gonna work.

The logic is like
![ketcher-review-1-2023-03-09-1343](https://user-images.githubusercontent.com/27288153/223933650-0b6bf21e-3556-404d-9ca5-09b7add324fb.png)

and your proposal is like
![ketcher-review-2-2023-03-09-1343](https://user-images.githubusercontent.com/27288153/223933842-eb436ea7-8778-4fef-9344-482f957ecc23.png)

They're two different logic. Anyway, you have brought up a good point, if you have any other ideas please don't hesitate to tell me! 😃